docs/architecture/adr-001-node-key-refactoring.md

```diff
@@ -0,0 +1,154 @@
+# ADR ADR-001: Node Key Refactoring
+
+## Changelog
+
+- 2022-10-31: First draft
+
+## Status
+
+Proposed
+
+## Context
+
+The original key format of IAVL nodes is a hash of the node. It does not take advantage of data locality on disk. Nodes are stored in random locations on the disk due to the random hash value, so it needs to scan the disk to find the corresponding node which can be very inefficient.
+
+The `orphans` are used to manage node removal in the current design and allow the deletion of removed nodes for the specific version from the disk through the `DeleteVersion` API. It needs to track every time when updating the tree and also requires extra storage to store `orphans`. But there are only 2 use cases for `DeleteVersion`:
+
+1. Rollback of the tree to a previous version
+2. Remove unnecessary old nodes
+
+## Decision
+
+- Use the version and the path as a node key like `bigendian(version) | byte array(path)` format. Here the `path` is a binary expression of the path from the root to the current node. 
+	```
+	`10101` : (right, left, right, left, right) -> [0x15]
+	```
+- Store only the child node key for the below version in node body writes. Because it is possible to get the child path for the same version.
+	```go
+	func (node *Node) getLeftNode() (*Node, error) {
+		if node.leftNode != nil {
+			return node.leftNode, nil
+		}
+		if node.leftNodeKey != nil {
+			return getNode(node.leftNodeKey) // get the node from the storage
+		}
+		return getNode(&NodeKey{
+			version: node.nodeKey.version,
+			path: node.nodeKey.path + '0',  // it will be more complicated in the real implementation
+		})
```

Comment:
Why are we getting a node with a new node key here? Doesn't this case mean that left is a leaf? I think we want to show how we're generating a new node key during insert/update

Reply:
this is just demonstrating how to skip `leftNodeKey` in node writes, without storing leftNodeKey, we can get the left node for the same version

Reply:
I see

Reply:
Does it make sense to at least store 1 bit of data as to whether the child is empty? This would prevent needing to do any sort of db read in that case

Reply:
what do you mean by the empty child? There are only two kind of nodes (inner node and leaf)
the `inner node` always have two child nodes